File: server/models/medicine.py
import uuid
from datetime import datetime, date, timezone
import enum

from sqlalchemy import String, DateTime, Date, ForeignKey, Enum as SAEnum, Boolean, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class MedicineFrequency(str, enum.Enum):
    ONCE_DAILY = "once_daily"
    TWICE_DAILY = "twice_daily"
    THREE_TIMES_DAILY = "three_times_daily"
    WEEKLY = "weekly"
    AS_NEEDED = "as_needed"
    CUSTOM = "custom"


class Medicine(Base):
    __tablename__ = "medicines"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    pet_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("pets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    name: Mapped[str] = mapped_column(String(120), nullable=False)
    dosage: Mapped[str] = mapped_column(String(80), nullable=False)
    frequency: Mapped[MedicineFrequency] = mapped_column(
        SAEnum(MedicineFrequency, name="medicine_frequency"), nullable=False
    )
    times_per_day: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    instructions: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    prescribed_by: Mapped[str | None] = mapped_column(String(120), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False
    )

    pet = relationship("Pet", backref="medicines")
    logs = relationship(
        "MedicineLog", backref="medicine", cascade="all, delete-orphan"
    )


class MedicineLog(Base):
    __tablename__ = "medicine_logs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    medicine_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("medicines.id", ondelete="CASCADE"), nullable=False, index=True
    )
    scheduled_for: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    given_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="pending", nullable=False)
    notes: Mapped[str | None] = mapped_column(String(500), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )

File: client/types/medicine.ts
export type MedicineFrequency =
  | "once_daily"
  | "twice_daily"
  | "three_times_daily"
  | "weekly"
  | "as_needed"
  | "custom";

export type MedicineLogStatus = "pending" | "given" | "missed";

export interface Medicine {
  id: string;
  petId: string;
  ownerId: string;
  name: string;
  dosage: string;
  frequency: MedicineFrequency;
  timesPerDay: number;
  startDate: string;
  endDate: string | null;
  instructions: string | null;
  prescribedBy: string | null;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface MedicineLog {
  id: string;
  medicineId: string;
  scheduledFor: string;
  givenAt: string | null;
  status: MedicineLogStatus;
  notes: string | null;
}

export interface MedicineFormPayload {
  petId: string;
  name: string;
  dosage: string;
  frequency: MedicineFrequency;
  timesPerDay: number;
  startDate: string;
  endDate?: string;
  instructions?: string;
  prescribedBy?: string;
  isActive: boolean;
}

interface RawMedicine {
  id: string;
  pet_id: string;
  owner_id: string;
  name: string;
  dosage: string;
  frequency: MedicineFrequency;
  times_per_day: number;
  start_date: string;
  end_date: string | null;
  instructions: string | null;
  prescribed_by: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

interface RawMedicineLog {
  id: string;
  medicine_id: string;
  scheduled_for: string;
  given_at: string | null;
  status: MedicineLogStatus;
  notes: string | null;
}

export function mapMedicine(raw: RawMedicine): Medicine {
  return {
    id: raw.id,
    petId: raw.pet_id,
    ownerId: raw.owner_id,
    name: raw.name,
    dosage: raw.dosage,
    frequency: raw.frequency,
    timesPerDay: raw.times_per_day,
    startDate: raw.start_date,
    endDate: raw.end_date,
    instructions: raw.instructions,
    prescribedBy: raw.prescribed_by,
    isActive: raw.is_active,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function mapMedicineLog(raw: RawMedicineLog): MedicineLog {
  return {
    id: raw.id,
    medicineId: raw.medicine_id,
    scheduledFor: raw.scheduled_for,
    givenAt: raw.given_at,
    status: raw.status,
    notes: raw.notes,
  };
}

export function toMedicinePayload(payload: MedicineFormPayload) {
  return {
    pet_id: payload.petId,
    name: payload.name,
    dosage: payload.dosage,
    frequency: payload.frequency,
    times_per_day: payload.timesPerDay,
    start_date: payload.startDate,
    end_date: payload.endDate || null,
    instructions: payload.instructions || null,
    prescribed_by: payload.prescribedBy || null,
    is_active: payload.isActive,
  };
}
